File: spicedb-gen/testdata/typescript/type_errors.ts
import { full } from "@spicedb/client";
import { TypedClient, Document, User, Team } from "./permissions";

declare const tc: TypedClient;

// @ts-expect-error: edit only reachable by user, not team#member
tc.check(full(), Document("x").edit, Team("eng").member());

// @ts-expect-error: delete only reachable by user
tc.check(full(), Document("x").delete, Team("eng").member());

// @ts-expect-error: editor relation only allows user
Document("x").editor(Team("eng"));

// @ts-expect-error: owner relation only allows user
Document("x").owner(Team("eng").member());

// @ts-expect-error: editor relation does not accept caveated user
Document("x").editor(User("alice").withIpRange({ allowedCidr: "10.0.0.0/8" }));

// @ts-expect-error: withIpRange requires a context object
User("alice").withIpRange();

// @ts-expect-error: wrong caveat context type (IpRangeContext has allowedCidr, not start/end)
User("alice").withIpRange({ start: "2024-01-01", end: "2024-12-31" });

File: spicedb-gen/testdata/python/type_errors.py
"""Deliberate type errors. pyright MUST reject this file.

Each call site is annotated with the expected error class.
"""

from __future__ import annotations

from permissions import (
    Document,
    Team,
    TeamMember,
    User,
    IpRangeContext,
    TimeWindowContext,
)

# 1. Team (the resource ref) is not a DocumentViewerSubject; only TeamMember is.
_ = Document("readme").viewer(Team("eng"))

# 2. TeamMember cannot be passed where DocumentEditorSubject (= User) is required.
_ = Document("readme").editor(TeamMember("eng"))

# 3. Wrong caveat-context type for with_ip_range.
_ = User("alice").with_ip_range(TimeWindowContext(start="x", end="y"))

# 4. Wrong field type on IpRangeContext.
_ = IpRangeContext(allowed_cidr=42)
